cpp/serve/engine_actions/new_request_prefill.cc

```diff
@@ -117,83 +120,130 @@ class NewRequestPrefillActionObj : public EngineActionObj {
           }
         }
         request_internal_ids.push_back(mstate->internal_id);
-        RECORD_EVENT(trace_recorder_, rsentry->request->id, "start embedding");
-        for (int i = 0; i < static_cast<int>(input_data.size()); ++i) {
-          embeddings = input_data[i]->GetEmbedding(models_[model_id],
-                                                   /*dst=*/!single_input ? &embeddings : nullptr,
-                                                   /*offset=*/cum_prefill_length);
-          cum_prefill_length += input_data[i]->GetLength();
+        RECORD_EVENT(trace_recorder_, prefill_inputs[i].rsentry->request->id, "start embedding");
+        // Speculative models shift left the input tokens by 1 when base model has committed tokens.
+        // Note: for n > 1 cases Eagle doesn't work because parent entry doesn't shift input tokens.
+        int embed_offset =
+            prefill_inputs[i].rsentry->mstates[model_id]->committed_tokens.empty() ? 0 : 1;
+        for (int j = 0; j < static_cast<int>(input_data.size()); ++j) {
+          if (j == static_cast<int>(input_data.size()) - 1) {
+            std::vector<int32_t> tail_tokens;
+            TokenData tk_data = Downcast<TokenData>(input_data[j]);
```

Comment:
We cannot assume it is `TokenData`. It's also possible to be ImageData.

Reply:
Is there any way to shift the input_data without downcasting to a specific type?

Reply:
If `input_data[-1]` is TokenData then it's fine to do such shift, or otherwise it remains a question whether shifting is feasible.

In this case, what about we use `const TokenData* tk_data = input_data[j].as<TokenDataNode>()` to get the pointer. use CHECK to ensure it is not nullptr, with the error message saying we eagle mode does not support multi-modality for now.